src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -556,29 +593,36 @@ impl Net {
                 self.rx_buffer
                     .iovec
                     .write_all_volatile_at(&self.rx_frame_buf[..vnet_hdr_len() + len], 0)?;
-                // SAFETY: This is safe:
+                // SAFETY:
+                // * len will never be bigger that u32::MAX
+                let len: u32 = (vnet_hdr_len() + len).try_into().unwrap();
+
+                // SAFETY:
                 // * We checked that `rx_buffer` includes at least one `DescriptorChain`
                 // * `rx_frame_buf` has size of `MAX_BUFFER_SIZE` and all `DescriptorChain` objects
                 //   are at least that big.
-                let dc = unsafe {
-                    self.rx_buffer
-                        .mark_used((vnet_hdr_len() + len).try_into().unwrap())
-                };
-
-                return Ok(Some(dc));
+                unsafe {
+                    self.rx_buffer.mark_used(len, &mut self.queues[RX_INDEX]);
+                }
+                return Ok(Some(len));
             }
         }
 
-        // SAFETY: this is safe because we ensured that `self.rx_buffer` has at least one
-        // DescriptorChain parsed in it.
+        // SAFETY:
+        // * We ensured that `self.rx_buffer` has at least one DescriptorChain parsed in it.
```

Comment:
nit (because unrelated to this PR): I don't think this needs to be unsafe. If we dont have any descriptor chains, we simply panic, or we pass an empty slice to `readv`. Both are fine and well-defined.